discovery: Add Go APM instrumentation detection

vitkyrka opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
-->
### What does this PR do?

Detect if a Go binary has APM instrumentation by checking for a symbol from dd-trace-go.  Only works on unstripped binaries.

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation

https://datadoghq.atlassian.net/browse/USMON-1191

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

vitkyrka committed:
discovery: Add Go APM instrumentation detection

Detect if a Go binary has APM instrumentation by checking for a symbol
from dd-trace-go.  Only works on unstripped binaries.

guyarb commented on pkg/collector/corechecks/servicediscovery/apm/detect.go:
```diff
@@ -85,6 +90,31 @@ func isInjected(envs map[string]string) bool {
+	}
+	defer elfFile.Close()
+
+	symbolsSet := common.StringSet{
+		"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer.init": struct{}{},
+	}
```
Should that be configurable and converted into a list?
is it possible future versions will have different names?

guyarb replied:
this code will break once ddtrace v2 will be released

vitkyrka replied:
Done.

vitkyrka committed:
bininspect: Add symbol filter inteface

Create an interface instead of the hard-coded StringSet filtering to
allow prefix-based symbol filtering.  In this commit, the external
interface is preserved.

vitkyrka committed:
bininspect: Make GetAllSymbolsByName() take SymbolFilter

Make GetAllSymbolsByName() take a SymbolFilter and make callers who need
StringSet based filtering use it explicitly.

vitkyrka committed:
bininspect: Add PrefixSymbolFilter

Add a SymbolFilter which finds symbols matching a prefix.

vitkyrka committed:
discovery: Make Go instrumentation detection more generic

vitkyrka committed:
Merge remote-tracking branch 'origin/main' into discovery-apm-go

 Conflicts:
	pkg/collector/corechecks/servicediscovery/apm/detect.go

vitkyrka committed:
bininspect: Make filter interface private

vitkyrka committed:
Merge remote-tracking branch 'origin/main' into discovery-apm-go

vitkyrka committed:
discovery: Fix typo in comment

guyarb commented on pkg/collector/corechecks/servicediscovery/apm/detect.go:
```diff
@@ -44,12 +46,14 @@ var (
+		language.Go:     goDetector,
 	}
 	// For now, only allow a subset of the above detectors to actually run.
 	allowedLangs = map[language.Language]struct{}{
 		language.Java:   {},
 		language.Node:   {},
 		language.Python: {},
+		language.Go:     {},
 	}
```
out of scope, but don't we just remove dotnet from dectectorMap?

guyarb commented on pkg/network/go/bininspect/symbol_filter.go:
```diff
@@ -0,0 +1,100 @@
+	return ok
+}
+
+func (f stringSetSymbolFilter) findMissing(symbolByName map[string]elf.Symbol) []string {
```
document

vitkyrka replied:
Done.

guyarb commented on pkg/network/go/bininspect/symbol_filter.go:
```diff
@@ -0,0 +1,100 @@
+}
+
+func (f stringSetSymbolFilter) findMissing(symbolByName map[string]elf.Symbol) []string {
+	missingSymbols := make([]string, 0, len(f.symbolSet)-len(symbolByName))
```
what guarantees `len(f.symbolSet) >= len(symbolByName)`?

guyarb replied:
since preallocation is an optimization, we should either use `max(0, len(f.symbolSet)-len(symbolByName))` or skip preallocation

vitkyrka replied:
Done.

guyarb commented on pkg/network/go/bininspect/symbol_filter.go:
```diff
@@ -0,0 +1,100 @@
+	return strings.HasPrefix(symbol, f.prefix)
+}
+
+func (f prefixSymbolFilter) findMissing(_ map[string]elf.Symbol) []string {
+	return []string{f.prefix + "..."}
+}
```
not sure I understand the implementation

vitkyrka replied:
I added a comment and dropped the .... It's only used for error prints.

guyarb commented on pkg/network/go/bininspect/symbols.go:
```diff
@@ -244,16 +248,36 @@ func GetAllSymbolsByName(elfFile *elf.File, symbolSet common.StringSet) (map[str
+	}
+
+	// Shouldn't happen
+	return nil, fmt.Errorf("empty symbols map")
```
```suggestion
	return nil, errors.New("empty symbols map")
```

vitkyrka committed:
discovery: Use errors.New

Co-authored-by: Guy Arbitman <guy20495@gmail.com>

vitkyrka committed:
discovery: Add bounds for len

vitkyrka committed:
discovery: Add comments

vitkyrka committed:
discovery: Only return prefix in findMissing

vitkyrka committed:
Merge remote-tracking branch 'origin/main' into discovery-apm-go

vitkyrka committed:
discovery: Use separate binary for Go instrumentation test

We added instrumentation to fake_server to test Go instrumentation but
this has resulted in some strange errors in KMT in CI which are not
reproducible locally in KMT.

We don't actually need to instrument fake_server but we just need some
binary which is instrumented so add a separate Go program which uses the
tracer but is never executed.

darccio commented:
Leaving here a comment so it's known: the import URLs used in this PR will be modified by adding a new set of URLs when we release `dd-trace-go/v2` (PR: https://github.com/DataDog/dd-trace-go/pull/2427).

guyarb commented:
@darccio please do not change the paths, we're using them in tests to verify we're able to detect go-instrumentation, instead we should be adding tests for `v2` to cover that as well
